### PR TITLE
fix: merge duplicate reply and mention notifications into one

### PR DESCRIPTION
**Pull request by llun**
## Problem

When someone replies to your post **and** mentions you in that reply, you receive **two** notifications for one event:

- "_<name> replied to your post_" (type `reply`)
- "_<name> mentioned you in a post_" (type `mention`)

This is the common case: a Mastodon-style reply auto-mentions the parent author, so almost every reply to you produced a duplicate. The two rows are created with different group keys (`reply:<parentId>` vs `mention:<statusId>`) and neither is in the default groupable set, so they never merge — in the UI **or** in the Mastodon API (both `reply` and `mention` serialise to the Mastodon `mention` type, so a Mastodon client also saw two "mention" notifications).

## Root cause

Two independent code paths each create a `reply` notification and a separate `mention` notification for the same recipient + status:

- `lib/services/timelines/mention.ts` — incoming **remote** replies.
- `lib/actions/createNote.ts` — **local** replies.

## Fix

Deduplicate at notification-creation time (the root cause), keeping the single, more specific **reply** notification and suppressing the duplicate **mention** one for the replied-to author:

- **Remote path** (`mention.ts`): once a reply notification is created for the recipient, skip creating the mention notification and its alert event for the same status.
- **Local path** (`createNote.ts`): skip the mention notification for the parent author when a reply notification is already created for them.

Why keep `reply` rather than `mention`:
- The native UI message "replied to your post" is more specific/useful than "mentioned you in a post".
- `reply` carries reply-specific email templates.
- `internalTypeToMastodon` maps `reply → mention`, so Mastodon API clients still see exactly one `mention`-type notification — the correct Mastodon behaviour (Mastodon has no separate reply type; a reply *is* a mention).

Actors who are mentioned but are **not** the replied-to author still receive normal `mention` notifications. Block/mute/filter policy behaviour is unchanged.

## Notes / scope

- This prevents **new** duplicates. Pre-existing duplicate rows already in the database are not retroactively cleaned up (no data migration) — that can be a follow-up if desired.
- No schema/migration changes.

## Tests

- Updated `lib/services/timelines/mention.test.ts`: the reply+mention case now asserts a single `reply` notification (no `mention` row) and a single reply alert event.
- Added `lib/actions/createNote.test.ts`: a local reply that also mentions the parent author creates exactly one (`reply`) notification and one reply alert.

## Verification

- `yarn run prettier --write .` ✓
- `yarn lint` ✓ (0 errors; only pre-existing warnings)
- `yarn build` ✓
- `yarn test` ✓ — 497 suites / 4261 tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/services/timelines/mention.ts [hunk 1]
```diff
@@ -75,8 +79,11 @@ export const mentionTimelineRule: MentionTimelineRule = async ({
...
               }
             )
-            if (replyNotification && !replyNotification.filtered) {
-              alertEvents.push({ type: NotificationType.enum.reply })
+            if (replyNotification) {
+              replyNotificationCreated = true
+              if (!replyNotification.filtered) {
+                alertEvents.push({ type: NotificationType.enum.reply })
+              }
             }
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/services/timelines/mention.ts]: Addressed in 1e4b4dd9b. `replyNotificationCreated` is now set as soon as the status is identified as a reply to the current actor — before inspecting the policy verdict — so the mention branch is suppressed consistently with the local `createNote.ts` path even when the reply notification is dropped/filtered, and the redundant mention policy evaluation is avoided.

Worth noting for accuracy: a *duplicate* mention could not actually have slipped through before, because `evaluateNotificationPolicy` returns the **same** verdict for `reply` and `mention` (the only type-aware dimension, `for_private_mentions`, treats `type === "mention" || type === "reply"` identically). So when the reply was dropped, the subsequent mention would have been dropped too — no extra row. The change is still correct and worthwhile: it aligns the two code paths, future-proofs against a type-divergent policy, and skips the redundant policy call. Added a filtered-reply regression test covering this branch.

[the remark this answers is not in the document]
Comment (llun) [reply lib/services/timelines/mention.ts]: Addressed in 1e4b4dd9b. `replyNotificationCreated` is now set as soon as the status is identified as a reply to the current actor — before inspecting the policy verdict — so the mention branch is suppressed consistently with the local `createNote.ts` path even when the reply notification is dropped/filtered, and the redundant mention policy evaluation is avoided.

Worth noting for accuracy: a *duplicate* mention could not actually have slipped through before, because `evaluateNotificationPolicy` returns the **same** verdict for `reply` and `mention` (the only type-aware dimension, `for_private_mentions`, treats `type === "mention" || type === "reply"` identically). So when the reply was dropped, the subsequent mention would have been dropped too — no extra row. The change is still correct and worthwhile: it aligns the two code paths, future-proofs against a type-divergent policy, and skips the redundant policy call. Added a filtered-reply regression test covering this branch.

**Comment by llun** [comment]
### 🤖 Sub-agent review — round 1

Ran a multi-lens adversarial review (correctness / edge-cases / tests / Mastodon-compat) with independent verification of each finding. **2 confirmed findings**, both now addressed in `1e4b4dd9b`:

1. **[correctness, high] Federation reply+mention silently dropped the email.** `lib/services/timelines/mention.ts` — the email payload was only ever attached to the *mention* event. Once the merge suppresses the mention branch, the surviving `reply` event carried no `emailContent`, so a remote reply-that-mentions-you sent **no email** when email is configured (push still fired). The local path (`createNote.ts`) already attaches a reply email; the federation path now does too. **Fix:** carry a reply email on the surviving reply event (added the `templates/reply` import, aliased the mention template). **Test:** asserts the merged reply event carries the reply email.

2. **[tests, medium] Filtered-reply branch was unasserted.** The merge sets `replyNotificationCreated` even when the reply is *filtered* (policy → requests queue), so a filtered reply+mention should collapse to one filtered `reply` notification with no mention and no alert — but nothing covered it. **Fix:** added a regression test (recipient policy `for_not_following: 'filter'`) asserting exactly one filtered `reply` notification, no `mention`, and no alert.

Findings that were raised but **rejected on verification** (not real for this diff): a claimed type-divergent policy bypass (the policy returns the same verdict for `reply` and `mention`, so no duplicate can occur) and a few style nits. Net raw findings: 6 → 2 confirmed after adversarial verification.

**Comment by llun** [comment]
### 🤖 Sub-agent review — round 2

Re-ran the multi-lens adversarial review against the updated diff (post email-fix). **1 confirmed finding** (coverage gap, not a present bug), addressed in `3f063934d`:

- **[tests, high] Untested boundary of the merge in `mention.ts`.** All merge tests had the recipient *be* the parent author. The boundary case — a remote reply to a **different** actor's post that also mentions the current actor — was uncovered. There the current actor isn't the parent, so `replyNotificationCreated` must stay `false` and the recipient must still get a `mention` notification with the **mention** email template. The code already does this correctly, but a regression that hoisted `replyNotificationCreated = true` out of the parent-identity guard would silently suppress third-party mentions (and swap the email template) with no failing test. **Fix:** added a test asserting a single `mention` notification + a mention event whose `emailContent.subject` matches the mention template (`mentions you in`), with no reply notification.

Raw findings 3 → 1 confirmed after verification. Gemini round 2: no feedback. Kilo: no issues.

**Comment by llun** [comment]
### 🤖 Sub-agent review — round 3 (converged)

Multi-lens adversarial review against the latest diff returned **0 findings** (0 raw). Combined with **Gemini round 3** ("no feedback to provide") and **Kilo** ("No Issues Found"), the review loop has converged — no new issues to address.

**Summary of the loop**

| Round | Gemini | Kilo | Sub-agents | Outcome |
|---|---|---|---|---|
| 1 | 1 (policy/flag consistency) | clean | 2 (email regression, filtered-test gap) | all fixed, threads replied + resolved |
| 2 | clean | clean | 1 (third-party-mention test gap) | test added |
| 3 | clean | clean | 0 | ✅ converged |

All commits: `6cc65a0d8` (fix) → `1e4b4dd9b` (reply email + flag consistency) → `3f063934d` (boundary test).

**CI:** all required checks green. One CI run had a flaky failure in `lib/utils/crypto.test.ts › generateAlphanumeric › has no obvious bias (chi-square test)` — an inherently probabilistic randomness test, unrelated to this change (it passed in the parallel run on the same commit and passes locally). Re-ran the failed jobs → green. Flagged separately as a follow-up to de-flake that test.
